domain/resource/state/resource.go

```diff
@@ -41,6 +42,162 @@ func NewState(factory database.TxnRunnerFactory, clock clock.Clock, logger logge
+
+	// SQL statement to list all resources for an application.
+	listAppResourcesStmt, err := st.Prepare(`
+SELECT resource_uuid AS &resourceIdentity.uuid 
+FROM application_resource 
+WHERE application_uuid = $resourceIdentity.application_uuid`, appIdentity)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// SQL statement to check there is no related resources in resource_file_store.
+	noFileStoreStmt, err := st.Prepare(`
+SELECT resource_uuid AS &resourceIdentity.uuid 
+FROM resource_file_store
+WHERE resource_uuid IN ($uuids[:])`, resourceIdentity{}, uuids{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// SQL statement to check there is no related resources in resource_image_store.
+	noImageStoreStmt, err := st.Prepare(`
+SELECT resource_uuid AS &resourceIdentity.uuid 
+FROM resource_image_store
+WHERE resource_uuid IN ($uuids[:])`, resourceIdentity{}, uuids{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// SQL statement to check there is no related resources in unit_resource.
+	noUnitResourceStmt, err := st.Prepare(`
+SELECT resource_uuid AS &resourceIdentity.uuid 
+FROM unit_resource
+WHERE resource_uuid IN ($uuids[:])`, resourceIdentity{}, uuids{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// SQL statement to delete resources from resource_retrieved_by.
+	deleteFromRetrievedByStmt, err := st.Prepare(`
+DELETE FROM resource_retrieved_by
+WHERE resource_uuid IN ($uuids[:])`, uuids{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// SQL statement to delete resources from application_resource.
+	deleteFromApplicationResourceStmt, err := st.Prepare(`
+DELETE FROM application_resource
+WHERE resource_uuid IN ($uuids[:])`, uuids{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// SQL statement to delete resources from resource.
+	deleteFromResourceStmt, err := st.Prepare(`
+DELETE FROM resource
+WHERE uuid IN ($uuids[:])`, uuids{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) (err error) {
+		// list all resources for an application.
+		var resources []resourceIdentity
+		err = tx.Query(ctx, listAppResourcesStmt, appIdentity).GetAll(&resources)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return err
+		}
+		resUUIDs := make(uuids, 0, len(resources))
+		for _, res := range resources {
+			resUUIDs = append(resUUIDs, res.UUID)
+		}
+
+		checkLink := func(message string, stmt *sqlair.Statement) error {
+			var resources []resourceIdentity
+			err := tx.Query(ctx, stmt, resUUIDs).GetAll(&resources)
+			switch {
+			case errors.Is(err, sqlair.ErrNoRows): // Happy path
+				return nil
+			case err != nil:
+				return err
+			}
+			return errors.Errorf("%s: %w", message, apperrors.InvalidCleanUpState)
+		}
+
+		// check there is no related resources in resource_file_store.
+		if err = checkLink("resource linked to file store data", noFileStoreStmt); err != nil {
```

Comment:
That a good remarks but we are checking non existence, not the possible existence. Their is no value to check or even document that it may or may be not linked due to some nature.

The intent is simple and clear: DONT DELETE WHAT IS STILL LINKED.

I think that not checking nor documenting what you suggest is the better way to enforce this intent. Sometimes, less is more.